### PR TITLE
Update flow to v0.54.0

### PR DESCRIPTION
**Pull request by puradox**
These changes solve the breaking changes of flow v0.53.0, which are
layed out [here](https://github.com/facebook/flow/blob/master/Changelog.md#0530).

Many of the changes were trivial, however there is one noticeable exception:
The props of the `Element` HOC has been changed to become all optional (I'm assuming this was the original intention), which in turn forced me to check if the props existing before using them in `_setupEventListeners()`.

Please double check that the changes in `src/components/Element.js` are correct before considering this PR, I wasn't quite sure about those changes.

### CODE REVIEW

File: demo/index.js [hunk 1]
```diff
@@ -48,7 +48,7 @@ const createOptions = (fontSize: string) => {
   };
 };
 
-class _CardForm extends React.Component {
```
Comment (michelle) [inline demo/index.js]: Hi! I think the right way to type this in 0.54 is to pull props (and state, if applicable) out from the _CardForm, like this:

```
type Props = {
  fontSize: string,
  stripe: StripeProps,
};

class _CardForm extends React.Component<Props> { ... }
```

File: src/components/Element.js [hunk 2]
```diff
@@ -1,22 +1,25 @@
...
-  elementRef: Function,
-  onChange: Function,
-  onBlur: Function,
-  onFocus: Function,
-  onReady: Function,
+  className?: string,
+  elementRef?: Function,
+  onChange?: Function,
+  onBlur?: Function,
+  onFocus?: Function,
```
Comment (michelle) [inline src/components/Element.js]: Hm, it surprises me that this is necessary, given we set default props below.

File: demo/index.js (hunk 1 above)
Comment (michelle) [reply demo/index.js]: Ah! I see you've done this below for the actual lib code :).

**Comment by michelle** [comment]
Thanks for the PR, @puradox! The change that necessitates checks for all of the event handlers confuses me a bit, because if we typed it as we did before w/ the right `defaultProps`, they should never be undefined or null, right?

**Comment by puradox** [comment]
@michelle Oh I must have missed `defaultProps` in the first run-through! You are correct, they should never be missing so there's no point of having them as optional. I'll make the appropriate changes.

**Comment by michelle-stripe** [comment]
Thanks! \o/

**Comment by puradox** [comment]
The problem seems more complicated than initially anticipated. Simply removing the optional flags on the Props of `Element` makes Flow freak out. 

It probably comes from the fact that `Element` really isn't a HOC, but it wants to act like it. 

> Concretely, a higher-order component is a function that takes a component and returns a new component.

Maybe we should consider changing `Element` to something like this:
```js
function makeElement(type: string, options: Options = {}): React.StatelessComponent<Props> {
  return (props) => <Element {...props} type={type} options={options} />
}
```

**Comment by michelle** [comment]
Sure, that seems reasonable to me. Though, `Element` components also need access to `context`.

**Comment by michelle** [comment]
I've not looked too deeply into it, but could this be related? https://github.com/facebook/flow/issues/4644#issue-251181403

**Comment by puradox** [comment]
Thanks @michelle for the link, that issue seems to be the problem we're running into here. I'm going to try the changes I mentioned above and see if that solves anything.

**Comment by puradox** [comment]
Sorry about the messy merge. It was not very fun.
Also, I'm getting eslint errors after the merge:
```
Cannot read config file: /home/sam/js/react-stripe-elements/node_modules/eslint-config-stripe/index.js
Error: Cannot find module 'eslint-config-airbnb-base'
Referenced from: /home/sam/js/react-stripe-elements/.eslintrc.yml
Error: Cannot read config file: /home/sam/js/react-stripe-elements/node_modules/eslint-config-stripe/index.js
Error: Cannot find module 'eslint-config-airbnb-base'
Referenced from: /home/sam/js/react-stripe-elements/.eslintrc.yml
```
When I add the additional devDependencies (which I didn't commit), it works fine. However, shouldn't `eslint-config-stripe` install those automatically?
Upon `yarn install` I get these warning:
```
warning "eslint-config-stripe@1.0.13" has incorrect peer dependency "eslint@^3.19.0".
warning "eslint-config-stripe@1.0.13" has incorrect peer dependency "eslint-plugin-import@^2.6.1".
warning "eslint-config-stripe@1.0.13" has incorrect peer dependency "eslint-plugin-jsx-a11y@^5.1.1".
warning "eslint-config-stripe@1.0.13" has incorrect peer dependency "eslint-plugin-react@^7.1.0".
warning "eslint-config-airbnb@15.1.0" has incorrect peer dependency "eslint-plugin-jsx-a11y@^5.1.1".
warning "eslint-config-airbnb@15.1.0" has incorrect peer dependency "eslint-plugin-import@^2.7.0".
warning "eslint-config-airbnb@15.1.0" has incorrect peer dependency "eslint-plugin-react@^7.1.0".
warning "eslint-config-airbnb-base@11.3.2" has incorrect peer dependency "eslint-plugin-import@^2.7.0".
```

**Comment by puradox** [comment]
Closing since #82 merged the necessary changes

**Comment by puradox** [comment]
Closing since #82 merged the necessary changes

**Pull request closed without merging**